### PR TITLE
Parser for the BiH company registry

### PR DESCRIPTION
**Pull request by dchaplinsky**
@jbothma please check the comments with `JD:` prefix.

### CODE REVIEW

File: datasets/ba/companies/crawler.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+        entity.add("address", record["address"], lang="bos")
+        if record["address_additional"]:
+            entity.add("address", record["address_additional"], lang="bos")
+
+        entity.add("legalForm", record["legal_form"], lang="bos")
+        entity.add("registrationNumber", record["registration_number"])
+
+        # JD: LegalEntity:idNumber probably?
+        if record.get("unique_id"):
+            entity.add("registrationNumber", record["unique_id"])
```
Comment (jbothma) [inline datasets/ba/companies/crawler.py:367]: damn, no support for Bosnia and Herzegovina yet in stdnum https://github.com/arthurdejong/python-stdnum/issues/268

yup, this one also looks like one `JIB	1234567890123	13 digits`

File: datasets/ba/companies/crawler.py [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+            director = context.make("Person")
+            director.id = context.make_id("BAdirector", entity.id, man["name"])
+            director.add("name", man["name"], lang="bos")
+            context.emit(director)
+
+            rel = context.make("Directorship")
+            rel.id = context.make_id("BADirectorship", entity.id, director.id)
+            rel.add("role", man["position"], lang="bos")
+            # JD: Should it be description or notes or status or summary?
+            rel.add("description", man["authorizations"], lang="bos")
```
Comment (jbothma) [inline datasets/ba/companies/crawler.py:425]: I would say description fits this well

File: datasets/ba/companies/crawler.py [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+
+            own = context.make("Ownership")
+            own.id = context.make_id("BAOwnership", entity.id, founder_company.id)
+            own.add("asset", entity.id)
+            own.add("owner", founder_company.id)
+            own.add("ownershipType", "Founder", lang="eng")
+
+            context.emit(own)
+
+        for man in managers:
```
Comment (jbothma) [inline datasets/ba/companies/crawler.py:415]: lol for funsies let's not make this man
```suggestion
        for manager in managers:
```
and then let python black do the formatting labour

File: datasets/ba/companies/crawler.py [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+            director = context.make("Person")
+            director.id = context.make_id("BAdirector", entity.id, man["name"])
+            director.add("name", man["name"], lang="bos")
+            context.emit(director)
+
+            rel = context.make("Directorship")
+            rel.id = context.make_id("BADirectorship", entity.id, director.id)
+            rel.add("role", man["position"], lang="bos")
```
Comment (jbothma) [inline datasets/ba/companies/crawler.py:423]: is position ever anything that might make us pick  a different schema? if so, we tend to do that using lookups, then there's no need to add it to :role.

File: datasets/ba/companies/crawler.py [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+    # JD: Jedinstveni identifikacioni broj - JIB or UIN
+    # https://www.vatify.eu/bosnia-and-herzegovina-vat-number.html
+    if uin:
+        record["unique_id"] = uin[0]
+
+    # JD: Don't know if we need this one
+    customs_number = details_page.xpath(
+        "//td[contains(text(),"
+        + ' "Customs Number")'
+        + "]/following-sibling::td/text()"
```
Comment (jbothma) [inline datasets/ba/companies/crawler.py:225]: I would chuck this into description as `f"Customs number: {customs_number}"`

File: datasets/ba/companies/crawler.py [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+
+    uin = details_page.xpath(
+        "//td[contains(text(),"
+        + ' "Unique Identification Number")'
+        + "]/following-sibling::td/text()"
+    )
+    # JD: Jedinstveni identifikacioni broj - JIB or UIN
+    # https://www.vatify.eu/bosnia-and-herzegovina-vat-number.html
+    if uin:
+        record["unique_id"] = uin[0]
```
Comment (jbothma) [inline datasets/ba/companies/crawler.py:219]: what's the question? or does this speak to the one where say yeah let's also add this as a second regno?

File: datasets/ba/companies/crawler.py [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+            context.emit(founder)
+
+            own = context.make("Ownership")
+            own.id = context.make_id("BAOwnership", entity.id, founder.id)
+            own.add("asset", entity.id)
+            own.add("owner", founder.id)
+            own.add("ownershipType", "Founder", lang="eng")
+
+            # JD: Not sure where to map shares and capital_contracted
+            own.add("sharesValue", person["capital_paid"])
```
[the remark this answers is not in the document]
Comment (jbothma) [reply datasets/ba/companies/crawler.py:387]: I don't understand capital contracted. But capital paid makes sense to add. Drop capital contracted unless you think it's really interesting, then tell me why.

**Review by jbothma — COMMENTED** [review]
Hey, this is looking great - I need to focus on something else but will come back tomorrow.

File: datasets/ba/companies/crawler.py [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+import datetime
+
+from zavod import Context
+
+
+# Unfortunatelly no cache for the listing page, as the state of the current
+# page is stored in the session and no cache for details page, as
+# the url is always changing
+CACHE_DAYS = 0
+SLEEP_TIME = 0.1  # seconds
```
[the remark this answers is not in the document]
Comment (dchaplinsky) [reply datasets/ba/companies/crawler.py:14]: I'm always trying to be nice.
Also, 429 won't really fix it, as we don't have an adaptive delays.
With 0.1 delay it'll take around 8-9 hours to scrape.

[the remark this answers is not in the document]
Comment (jbothma) [reply datasets/ba/companies/crawler.py:14]: I mean I'd bet money that they don't send the Retry-After header, but we use https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html#urllib3.util.Retry under the hood which respects it.

File: datasets/ba/companies/crawler.py [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+                {
+                    "basic_data": bd,
+                    "registration_number": (
+                        reg_num[i] if re.search(r"\d", reg_num[i]) else ""
+                    ),
+                    "name": company_name,
+                    "address": address,
+                    "country": country,
+                }
+            )
```
[the remark this answers is not in the document]
Comment (jbothma) [reply datasets/ba/companies/crawler.py:311]: Ah I get it. So you get prefixes here that aren't company names? Could you see if those prefixes are sufficiently regular that we can strip them out (until a new unexpected value pops up)?

Screenshot 2024-04-17 at 14 08 04

**Review by jbothma — COMMENTED** [review]
overall super nice. thanks!

File: datasets/ba/companies/crawler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dchaplinsky) [reply datasets/ba/companies/crawler.py:14]: Should I set it to zero then?

File: datasets/ba/companies/crawler.py (hunk 6 above)
Comment (dchaplinsky) [reply datasets/ba/companies/crawler.py:219]: The question is in the another comment :)

File: datasets/ba/companies/crawler.py (hunk 4 above)
Comment (dchaplinsky) [reply datasets/ba/companies/crawler.py:421]: Sorry, I don't understand this part and how it works. `lookups` is not really described in the docs, I fear.

File: datasets/ba/companies/crawler.py (hunk 1 above)
Comment (dchaplinsky) [reply datasets/ba/companies/crawler.py:365]: Ok, what should I do? Ignore?

Comment (jbothma) [reply datasets/ba/companies/crawler.py:346]: Nope, this is perfect

**Comment by jbothma** [comment]
Дякую

**Comment by dchaplinsky** [comment]
My pleasure!

**Pull request merged**